superwall/src/main/java/com/superwall/sdk/web/WebPaywallRedeemer.kt

```diff
@@ -0,0 +1,350 @@
+    val maxAge: () -> Long,
+    val getActiveDeviceEntitlements: () -> Set<Entitlement> = {
+        Superwall.instance.entitlements.activeDeviceEntitlements
+    },
+    val getUserId: () -> UserId? = {
+        Superwall.instance.dependencyContainer.identityManager.appUserId?.let {
+            UserId(
+                it,
+            )
+        }
+    },
+    val getDeviceId: () -> DeviceVendorId = { DeviceVendorId(Superwall.instance.vendorId) },
+    val getAliasId: () -> String? = { Superwall.instance.dependencyContainer.identityManager.aliasId },
+    val track: suspend (Trackable) -> Unit = {
+        Superwall.instance.track(it)
+    },
+    val internallySetSubscriptionStatus: (SubscriptionStatus) -> Unit,
+    val isPaywallVisible: suspend () -> Boolean = { Superwall.instance.isPaywallPresented },
+    val triggerRestoreInPaywall: suspend () -> Unit = { },
+    val currentPaywallEntitlements: () -> Set<Entitlement>,
+    val getPaywallInfo: () -> PaywallInfo,
+    val trackRestorationFailed: (message: String) -> Unit,
+    val isWebToAppEnabled: () -> Boolean,
+) {
+    private var pollingJob: Job? = null
+
+    init {
+        ioScope.launch {
+            checkForRefferal()
+            startPolling()
+        }
+    }
+
+    suspend fun checkForRefferal() =
+        asEither {
+            if (isWebToAppEnabled()) {
+                deepLinkReferrer
+                    .checkForReferral()
+                    .fold(
+                        onSuccess = {
+                            redeem(RedeemType.Code(it))
+                        },
+                        onFailure = { throw it },
+                    )
+            }
+        }
+
+    sealed interface RedeemType {
+        data class Code(
+            val code: String,
+        ) : RedeemType
+
+        object Existing : RedeemType
+    }
+
+    suspend fun redeem(redemption: RedeemType) {
+        // We want to keep track of the codes that have been retrieved by the user
+        val latestResponse = storage.read(LatestRedemptionResponse)
+        val allCodes = latestResponse?.allCodes?.toMutableList() ?: mutableListOf()
+        var isFirstRedemption = true
+        if (redemption is RedeemType.Code) {
+            willRedeemLink()
+
+            if (allCodes.isNotEmpty()) {
+                isFirstRedemption = !allCodes.map { it.code }.contains(redemption.code)
+            }
+            allCodes.add(Redeemable(redemption.code, isFirstRedemption))
+
+            if (isPaywallVisible()) {
+                track(
+                    InternalSuperwallEvent.Restore(
+                        InternalSuperwallEvent.Restore.State.Start,
+                        getPaywallInfo(),
+                    ),
+                )
+            }
+            track(Redemptions(RedemptionState.Start))
+        }
+        network
+            .redeemToken(
+                allCodes,
+                getUserId(),
+                getAliasId(),
+                getDeviceId(),
```

Comment:
On iOS we're also sending the transaction receipts here, a jws representation of the first transaction for each subscription. This might not apply to android though